Support Google Cloud Storage in addition to S3

mythmon opened the pull request:
Fixes #1566

sciurus commented on normandy/base/storage.py:
```diff
@@ -30,9 +31,10 @@ def get_valid_name(self, name):
 
         name = force_text(name).strip()
-        # remove "characters to avoid", as described by S3's docs
+        # remove "characters to avoid", as described by S3's GCP's docs
```
Nit: _S3's and GCP's_

sciurus commented on normandy/base/storage.py:
```diff
@@ -30,9 +31,10 @@ def get_valid_name(self, name):
         # https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-key-guidelines-avoid-characters
-        name = re.sub(r'[\\{}^%`\[\]<>~#|\x00-\x1F\x7F-\xFF\'"]', "", name)
+        # https://cloud.google.com/storage/docs/naming#objectnames
+        name = re.sub(r'[\\{}^%`\[\]<>~#|\x00-\x1F\x7F-\xFF\'"*?]', "", name)
```
Our of curiosity and paranoia, I manually tested this successfully.

peterbe commented:
Can you merge master? @mythmon

peterbe commented on normandy/base/storage.py:
```diff
@@ -6,20 +6,21 @@
     """
     A storage class mixin that is is more permissive about valid filenames.
 
     In the original implemtnation, "anything that is not a unicode
```
might as well fix this "implemtnation" typo :)

peterbe commented on normandy/base/storage.py:
```diff
@@ -6,20 +6,21 @@
-    Characters that S3 describes as requiring "significant special
-    handling for consistency across all applications" are removed.
+    File names produced follow recommended guidelines for both AWS S3 and GCP
+    Cloud Storage object names.
```
nit: If you have a URL to that "recommended guidelines" that would be great.

mythmon replied:
The URL is linked in the code, next to the actual changes to follow the guidelines. Does that work?

mythmon commented:
I fixed the nits and rebased on master to fix merge conflicts.

> I have not tested this on a real GCS account, have you?

I have not tested this directly either. I haven't worked on that kind of set up yet. I'm comfortable merging this without that testing though, since we aren't deploying anything just yet.

peterbe commented:
I've actually never tested any real S3 uploads from localhost myself. That's how far behind I am :)

Next up on my personal todo list is to get comfortable with GCS for the sake of Tecken so hopefully I can leverage some of that experience in normandy/DC and test for realz.

mythmon commented:
Nits are fixed, and I have approvals. Removing the label and merging this.

bors r= @peterbe, @sciurus
